test/l1/L1DAIWormholeBridge.ts

```diff
@@ -0,0 +1,129 @@
+import { ethers } from 'hardhat'
+
+import { Dai__factory, L1DAIWormholeBridge__factory, L1Escrow__factory } from '../../typechain-types'
+import { deployAbstractMock, deployMock, deployOptimismContractMock } from '../helpers'
+
+const INITIAL_ESCROW_BALANCE = 3000
+const SOURCE_DOMAIN_NAME = ethers.utils.formatBytes32String('optimism-a')
+const TARGET_DOMAIN_NAME = ethers.utils.formatBytes32String('arbitrum-a')
+const AMOUNT = 100
+
+describe('L1DAIWormholeBridge', () => {
+  it('has correct public interface', async () => {
+    await assertPublicMutableMethods('L1DAIWormholeBridge', [
+      'finalizeFlush(bytes32,uint256)',
+      'finalizeRegisterWormhole((bytes32,bytes32,address,address,uint128,uint80,uint48))',
+    ])
+  })
+
+  describe('constructor', () => {
+    it('assigns all variables properly', async () => {
+      const [l2DAIWormholeBridge, l1CrossDomainMessenger, l1Escrow, wormholeRouter] = await ethers.getSigners()
+
+      const l1Dai = await simpleDeploy<Dai__factory>('Dai', [])
+      const l1DAITokenBridge = await simpleDeploy<L1DAIWormholeBridge__factory>('L1DAIWormholeBridge', [
+        l1Dai.address,
+        l2DAIWormholeBridge.address,
+        l1CrossDomainMessenger.address,
+        l1Escrow.address,
+        wormholeRouter.address,
+      ])
+
+      expect(await l1DAITokenBridge.l1Token()).to.eq(l1Dai.address)
+      expect(await l1DAITokenBridge.l2DAIWormholeBridge()).to.eq(l2DAIWormholeBridge.address)
+      expect(await l1DAITokenBridge.escrow()).to.eq(l1Escrow.address)
+      expect(await l1DAITokenBridge.messenger()).to.eq(l1CrossDomainMessenger.address)
+      expect(await l1DAITokenBridge.wormholeRouter()).to.eq(wormholeRouter.address)
+    })
+  })
+
+  describe('finalizeFlush', () => {
+    it('calls the router to settle the dai debt', async () => {
+      const [_, l1MessengerImpersonator, routerImpersonator, wormholeJoin] = await ethers.getSigners()
+      const {
+        l1Dai,
+        l1DAIWormholeBridge,
+        l1CrossDomainMessengerMock,
+        l2DAIWormholeBridge,
+        l1Escrow,
+        wormholeRouterMock,
+      } = await setupTest({ l1MessengerImpersonator, routerImpersonator })
+      l1CrossDomainMessengerMock.smocked.xDomainMessageSender.will.return.with(() => l2DAIWormholeBridge.address)
+
+      await waitForTx(l1DAIWormholeBridge.connect(l1MessengerImpersonator).finalizeFlush(TARGET_DOMAIN_NAME, AMOUNT))
+      const routerSettleCallData = wormholeRouterMock.smocked.settle.calls[0]
+      await waitForTx(
+        // mock DAI transfer by router
+        l1Dai.connect(routerImpersonator).transferFrom(l1DAIWormholeBridge.address, wormholeJoin.address, AMOUNT),
+      )
+
+      expect(routerSettleCallData.targetDomain).to.equal(TARGET_DOMAIN_NAME)
+      expect(routerSettleCallData.batchedDaiToFlush).to.equal(AMOUNT)
+      expect(await l1Dai.balanceOf(l1Escrow.address)).to.eq(INITIAL_ESCROW_BALANCE - AMOUNT)
+      expect(await l1Dai.balanceOf(wormholeJoin.address)).to.eq(AMOUNT)
```

Comment:
can we add a test that this should fail if called not via xdomain messagenger or called by wrong l2 control? (unexpected xDomainMessageSender)